umbraco/Umbraco-CMS
# Examine 2.0 integration

WIP. This is integrating Examine 2.0 (netstandard) into Umbraco v9. 

## Examine 2.0 changes

* 2.0.0-beta.136 is on MyGet
* All index configuration is done during composition. This is different to how it behaved in v8 but is a much better and will end up being simpler for end users.
* The same Umbraco Examine configuration options are still supported.
* Customizing an index can now easily be done using [IPostConfigureOptions](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.options.ipostconfigureoptions-1.postconfigure?view=dotnet-plat-ext-5.0). Examine 2.0 uses named IOptions which is convenient because each index is created with a name. So registering something like this in DI will allow end users to customize the indexes. The options contain field definitions, directory factory, analyzer, validator, etc... Umbraco itself customizes the indexes this way but uses IConfigureNamedOptions. It would be possible for end users to use that too but would need to make sure it's registered after Umbraco's registrations. This is why it's easier to just use IPostConfigureOptions since we know it will execute after.
   ```cs
    public sealed class ConfigureIndexOptions : IPostConfigureOptions<LuceneDirectoryIndexOptions>
    {
        public void PostConfigure(string name, LuceneDirectoryIndexOptions options)
        {
            switch (name)
            {
                case "ExternalIndex":
                    options.Analyzer = new StandardAnalyzer(LuceneInfo.CurrentVersion);
                    break;
            }
        }
    }
   ```
* Lucene indexes are unlocked automatically by Examine itself so we can remove any custom Umbraco code that was doing that before
* Fixes an issue with the Umbraco Examine implementation where we were double validating content entries. This should perform better.
* Moves and fixes up the Umbraco Examine unit tests. There are some tests for Umbraco Examine still buried in the old published content tests but until published content tests are moved to net 5 we can't move those.
* Adjust some namespaces, we have `Umbraco.Cms.Infrastructure.Examine` which is all to do with Umbraco Examine and we also have `Umbraco.Cms.Infrastructure.Search` which is generic search operations in Umbraco. 
* refactors `IIndexRebuilder`, no more background rebuilder, etc... there's just one 
* new `IUmbracoIndexingHandler` which is an abstraction for handling index operations triggered by Umbraco. Umbraco will listen for all required changes to it's data structures and then call methods on `IUmbracoIndexingHandler` to do the indexing. This means that if you wanted to replace indexing by Umbraco Examine with something else you can just replace this one interface.
* Rebuilding on startup is done with handler that will start the process when the first HttpRequest is started with a delay of 1 minute.
* Examine 2.0 has some slightly simpler APIs especially with resolving a searcher. Now you just do `index.Searcher` instead of `index.GetSearcher()`.  SkipTake is also a simpler API for paging. 
* 

## Other Changes

There are other supporting changes:

* Umbraco.Tests.Common targets net5 not netstandard
* This integrates #10182 from v8 into v9, though since we are allowed breaking changes this does it in a nicer way.
  * This includes a few new things: `ISyncBootStateAccessor` to be able to inject in various services to know if it's a ColdBoot or WarmBoot. 
  * No more DatabaseServerMessengerCallbacks. These were also removed from the v8 PR but we can completely delete them here
  * New LastSyncedFileManager used to manage access to the lastsynced.txt file
  * DatabaseServerMessenger updated 
* Ensures all INotifications are suffixed with the term "Notification"
* Removes MD5 hashing (this is a legacy hashing algorithm and shouldn't be used)
* CacheInstructionService changed to not inject CacheRefresherCollection/ServerRole because that results in circular deps. We pass these params into the method now.
* Removes TaskHelper (moves to test proj for now since it's still used there) - background tasks initiated will use the new IBackgroundTaskQueue which is a dedicated background task queue and doesn't have an issue of leaking ExecutionContext because it's a pre-launched thread.
* Moves `UmbracoApplicationStartingNotification` to be published after all components instead of before.
* Fixes `IHostingEnvironment.ApplicationId`. Previously this was returning the AppDomain (INT) Id which is incorrect. This needs to be a unique identifier for the application and be the same between restarts. In v8 this was the IIS Id as resolved by `HttpRuntime.AppDomainAppId` which would be a value like `/LM/W3C/<site id>/<servicename>`. In aspnetcore there is no 100% equivalent value but there is a public API exposed for it. For more info see: https://stackoverflow.com/a/67329922 . The `IHostingEnvironment.ApplicationId` is used for various things: MainDom (having it as the AppDomain INT id will totally break MainDom), `ServerRegistrationService` for storing the worker's unique ID in the database (this service isn't really used anymore and should be overhauled in a better way so that we can store the lastsynced.txt file in the DB, see https://github.com/umbraco/Umbraco-CMS/issues/5263), DatabaseServerMessenger to be able to identify the 'current' worker, LastSyncedFileManager in order to saved the lastsynced.txt file based on the unique identifier of the worker.
* No  more NetworkHelper, changes to EnvironmentHelper and makes internal
* MainDom.Register is safer to use, previously we would just log a warning if Register was called before acquiring took place but that is an illegal operation and would cause problems. So now we throw when that scenario occurs. This issue was noticed because PublishedSnapshotService was never reading from the local cache files and was always cold booting because it was registering with MainDom in it's ctor which was executed before MainDom was acquired. So now PublishedSnapshotService registers with MainDom lazily in it's EnsureCaches call

## Known issues

* TODO: we need to test when hosted on IIS if the HttpRuntime.AppDomainAppId resolves to the previous IIS unique identifier. 
* TODO: the implementation in v8 for lastsynced.txt is incorrect and must also contain the machine name
* If there are no indexes for a site and a user navigates to the examine dashboard before the ExamineIndexRebuilder.RebuildIndexes timeout has elapsed, then the indexes will be created but rebuilding and populating the indexes with the ExamineIndexRebuilder.RebuildIndexes will not occur because it will detect that the indexes exist. The only way I can see working around this would be too have a check in the ExamineManagementController.GetIndexerDetails that checks if indexes don't yet exist and then initiate and index rebuild/populate immediately. This would however require further UI work to inform the user of what is happening since index rebuilding is not a blocking operation and it will be triggered in the background. 

## Testing

* Probably best to delete all indexes first (/Umbraco/Data/TEMP/ExamineIndexes)
* Ensure cold boot works (delete all files in /Umbraco/Data/TEMP/DistCache)
* Ensure index are rebuilt on startup - you will need to wait 1 minute after startup for this to occur
* Ensure the examine mgmt dashboard works and you can rebuild indexes there
* Ensure that saving content, media and members populates the indexes
* All tests should be passing
* Run some testing on the above changes
